ios/viewmodels/view-model-reuse.md

```diff
@@ -0,0 +1,167 @@
+
+![ProfileViewController](profile.png)
+![ReviewViewController](review.png)
+
+Let's call these two views ProfileViewController and ReviewViewController respectively.
+
+The table view cell pictured here is the same in both cases, and we would ideally like to reuse the table view cell and it's xib. In both cases, we are trying to abstract view logic out of the table view cell. However, the models that drive these two cells are very different. Let's explore two approaches of view models that can be used for this view.
+
+### Concrete view model class
+
+This simpler of the two approaches just has a concrete class for the view, which is how we would do it if the cell only had to display one kind of model.
+
+```swift
+
+struct TitleMessageCellViewModel {
+  let title: String
+  let message: String
+  let showsEditButton = false
+}
+
+class TitleMessageCell: UITableViewCell {
+  @IBOutlet private weak var titleLabel: UILabel!
+  @IBOutlet private weak var messageLabel: UILabel!
+
+  private func configure(viewModel: TitleMessageCellViewModel) {
+    titleLabel.text = viewModel.title.uppercaseString
+    messageLabel.text = viewModel.message
+    editButton.hidden = !viewModel.showsEditButton
+  }
+}
+
+```
+
+An implementation for ProfileViewController might look like this:
+
+```swift
+
+// ProfileViewController.swift
+
+func viewModel(forIndex index: Int) {
+  switch index {
+    case 0:
+      return TitleMessageCellViewModel(title: "First Name", message: profile.firstName)
+    case 1:
+      return TitleMessageCellViewModel(title: "Surname", message: profile.lastName)
+    ...
+  }
+}
+
+```
+
+This implementation abstracts out the view logic details out of the table view cell, albeit moving it to the view controller. When the view logic is trivial, this is not too bad.
+
+The implementation for ReviewViewController, however, might look like this:
+
+```swift
+
+// ReviewViewController.swift
+
+func viewModel(forIndex index: Int) {
+  switch index {
+    case 0:
+      let formatter = PhoneNumberFormatter(forCountryCode: currentCountryCode)
+      let number = formatter.formatNumber(profile.phoneNumber) ?? "-"
+      return TitleMessageCellViewModel(title: "Contact number", message: number, showsEditButton: profile.canEdit)
+    case 1:
+      let rows = topics.filter({ $0.selected }).map({ $0.title }).joinWithSeparator("\n")
+      return TitleMessageCellViewModel(title: "Visit Details", message: topics, showsEditButton: review.canEdit)
+    ...
+  }
+}
+
+```
+
+The problem with this approach is that when the view logic is trivial (ProfileViewController), we can get away with implementing view logic within the view controller, but as soon as it gets complicated(ReviewViewController), the view controller starts knowing a lot about view logic details, and we are back to view controller and view bloat.
```

Comment:
I'm saying that this statement:

> The problem with this approach is ... as soon as it gets complicated(ReviewViewController), the view controller starts knowing a lot about view logic details, and we are back to view controller and view bloat

Is untrue.  At least, it is untrue for the approach I assume you are talking about -- using a concrete VM class instead of a protocol based VM.

The *real* cause of the VC/V bloat is putting the VM creation logic into the VC/V.  When you used a concrete VM class you only *exposed* the issue.  You didn't *cause* it.

And it isn't corrected below -- you still create the VM from the VC, not from the parent VM.  So any configuration logic that needs to happen externally to that VM is still going to have to happen in the VC.  Yes, you were able to encapsulate that logic by implementing classes with very specific roles, but you still need to make that decision of which to use inside the VC.

Take this pretty far -- say you have a lot of configuration options that need to be made at creation time.  You end up with all that logic that decides which class to use inside the VC, which continues to bloat the VC.  Instead, that logic should live in the VM (assuming it is ultimately based on model states) and not bloat the VC.

Another way to think about it: Which VM to use is not actually dependent on anything view-related.  It isn't because I tapped on a particular section of the screen.  It isn't because the height of the cell is 50pt.  It isn't because there are two labels in the cell.  It is because of what *the backing model behind the cell* is.  Thus, it should be the parent VM's job to determine which cell VM to use.  The parent VM has all the context of what model is backing the cell.

So I think to summarize: I think the premise you lay out here is untrue.  I think the example is fine, but what you are *actually* solving by using protocol based VMs is not exactly what you say it is.  What you say you are solving is actually a symptom of something else.  What you actually solve is the general issue of bloated classes (classes doing too many things), not bloated VC/Vs in particular.